File: src/lib/services/notification.ts
import nodemailer from "nodemailer";
import type { IEndpoint } from "@/lib/models/endpoint";
import type { CheckResultData } from "@/types";
import Notification from "@/lib/models/notification";

const RATE_LIMIT_MS = 5 * 60 * 1000; // 5 minutes

export async function processNotifications(
  endpoint: IEndpoint,
  result: CheckResultData
): Promise<void> {
  if (!endpoint.alertEnabled) return;

  const wasAlerting = endpoint.isAlerting;
  const isNowUp = result.status === "UP";

  // Recovery notification
  if (wasAlerting && isNowUp) {
    await sendToChannels(endpoint, "recovery", result);
    endpoint.isAlerting = false;
    await endpoint.save();
    return;
  }

  // Alert notification
  if (
    endpoint.consecutiveFailures >= endpoint.alertThreshold &&
    !wasAlerting
  ) {
    // Rate limit check
    if (
      endpoint.lastAlertedAt &&
      Date.now() - endpoint.lastAlertedAt.getTime() < RATE_LIMIT_MS
    ) {
      return;
    }

    await sendToChannels(endpoint, "alert", result);
    endpoint.isAlerting = true;
    endpoint.lastAlertedAt = new Date();
    await endpoint.save();
  }
}

async function sendToChannels(
  endpoint: IEndpoint,
  type: "alert" | "recovery",
  result: CheckResultData
): Promise<void> {
  const message =
    type === "alert"
      ? `[ALERT] ${endpoint.name} is ${result.status} (${endpoint.consecutiveFailures} consecutive failures)`
      : `[RECOVERED] ${endpoint.name} is back UP`;

  const promises: Promise<void>[] = [];

  if (endpoint.notifications?.email?.enabled && endpoint.notifications.email.address) {
    promises.push(
      sendEmail(endpoint.notifications.email.address, message, endpoint, type)
    );
  }

  if (endpoint.notifications?.discord?.enabled && endpoint.notifications.discord.webhookUrl) {
    promises.push(
      sendDiscord(endpoint.notifications.discord.webhookUrl, message, endpoint, result, type)
    );
  }

  if (endpoint.notifications?.webhook?.enabled && endpoint.notifications.webhook.url) {
    promises.push(
      sendWebhook(endpoint.notifications.webhook.url, endpoint, result, type)
    );
  }

  await Promise.allSettled(promises);
}

async function sendEmail(
  to: string,
  message: string,
  endpoint: IEndpoint,
  type: "alert" | "recovery"
): Promise<void> {
  try {
    const transporter = nodemailer.createTransport({
      host: process.env.SMTP_HOST,
      port: parseInt(process.env.SMTP_PORT || "587"),
      auth: {
        user: process.env.SMTP_USER,
        pass: process.env.SMTP_PASS,
      },
    });

    await transporter.sendMail({
      from: process.env.SMTP_FROM || "Pulse <noreply@pulse.dev>",
      to,
      subject: message,
      text: `Endpoint: ${endpoint.name}\nURL: ${endpoint.url}\nStatus: ${endpoint.lastStatus}\nTime: ${new Date().toISOString()}`,
    });

    await logNotification(endpoint, "email", type, "sent", message);
  } catch (err) {
    await logNotification(
      endpoint,
      "email",
      type,
      "failed",
      message,
      (err as Error).message
    );
  }
}

async function sendDiscord(
  webhookUrl: string,
  message: string,
  endpoint: IEndpoint,
  result: CheckResultData,
  type: "alert" | "recovery"
): Promise<void> {
  const color = type === "alert" ? 0xff0000 : 0x00ff00;
  const payload = {
    embeds: [
      {
        title: message,
        color,
        fields: [
          { name: "Endpoint", value: endpoint.name, inline: true },
          { name: "URL", value: endpoint.url, inline: true },
          { name: "Status", value: result.status, inline: true },
          ...(type === "alert"
            ? [
                {
                  name: "Consecutive Failures",
                  value: String(endpoint.consecutiveFailures),
                  inline: true,
                },
              ]
            : []),
        ],
        timestamp: new Date().toISOString(),
      },
    ],
  };

  try {
    await fetch(webhookUrl, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });

    await logNotification(endpoint, "discord", type, "sent", message);
  } catch (err) {
    await logNotification(
      endpoint,
      "discord",
      type,
      "failed",
      message,
      (err as Error).message
    );
  }
}

async function sendWebhook(
  url: string,
  endpoint: IEndpoint,
  result: CheckResultData,
  type: "alert" | "recovery"
): Promise<void> {
  const payload = {
    endpoint: endpoint.name,
    url: endpoint.url,
    status: result.status,
    consecutiveFailures: endpoint.consecutiveFailures,
    checkedAt: new Date().toISOString(),
    type,
  };

  try {
    await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });

    await logNotification(
      endpoint,
      "webhook",
      type,
      "sent",
      JSON.stringify(payload)
    );
  } catch (err) {
    await logNotification(
      endpoint,
      "webhook",
      type,
      "failed",
      JSON.stringify(payload),
      (err as Error).message
    );
  }
}

async function logNotification(
  endpoint: IEndpoint,
  channel: "email" | "discord" | "webhook",
  type: "alert" | "recovery",
  status: "sent" | "failed",
  message: string,
  error?: string
): Promise<void> {
  await Notification.create({
    endpointId: endpoint._id,
    channel,
    type,
    status,
    message,
    error: error || null,
  });
}

File: backend/services/notification.py
from datetime import datetime, timezone

import httpx

from core.config import settings
from models.endpoint import Endpoint
from models.notification import Notification

RATE_LIMIT_SECONDS = 5 * 60  # 5 minutes


async def process_notifications(endpoint: Endpoint, result: dict) -> None:
    if not endpoint.alert_enabled:
        return

    was_alerting = endpoint.is_alerting
    is_now_up = result["status"] == "UP"

    # Recovery
    if was_alerting and is_now_up:
        await send_to_channels(endpoint, "recovery", result)
        endpoint.is_alerting = False
        await endpoint.save()
        return

    # Alert
    if endpoint.consecutive_failures >= endpoint.alert_threshold and not was_alerting:
        if endpoint.last_alerted_at:
            elapsed = (datetime.now(timezone.utc) - endpoint.last_alerted_at).total_seconds()
            if elapsed < RATE_LIMIT_SECONDS:
                return

        await send_to_channels(endpoint, "alert", result)
        endpoint.is_alerting = True
        endpoint.last_alerted_at = datetime.now(timezone.utc)
        await endpoint.save()


async def send_to_channels(endpoint: Endpoint, notif_type: str, result: dict) -> None:
    message = (
        f"[ALERT] {endpoint.name} is {result['status']} ({endpoint.consecutive_failures} consecutive failures)"
        if notif_type == "alert"
        else f"[RECOVERED] {endpoint.name} is back UP"
    )

    if endpoint.notifications.email.enabled and endpoint.notifications.email.address:
        await send_email(endpoint.notifications.email.address, message, endpoint, notif_type)

    if endpoint.notifications.discord.enabled and endpoint.notifications.discord.webhook_url:
        await send_discord(endpoint.notifications.discord.webhook_url, message, endpoint, result, notif_type)

    if endpoint.notifications.webhook.enabled and endpoint.notifications.webhook.url:
        await send_webhook(endpoint.notifications.webhook.url, endpoint, result, notif_type)


async def send_email(to: str, message: str, endpoint: Endpoint, notif_type: str) -> None:
    try:
        if not settings.SMTP_HOST:
            await log_notification(endpoint, "email", notif_type, "failed", message, "SMTP not configured")
            return

        import aiosmtplib
        from email.mime.text import MIMEText

        msg = MIMEText(
            f"Endpoint: {endpoint.name}\nURL: {endpoint.url}\n"
            f"Status: {endpoint.last_status}\nTime: {datetime.now(timezone.utc).isoformat()}"
        )
        msg["Subject"] = message
        msg["From"] = settings.SMTP_FROM
        msg["To"] = to

        await aiosmtplib.send(
            msg,
            hostname=settings.SMTP_HOST,
            port=settings.SMTP_PORT,
            username=settings.SMTP_USER or None,
            password=settings.SMTP_PASS or None,
            start_tls=True,
        )
        await log_notification(endpoint, "email", notif_type, "sent", message)
    except Exception as exc:
        await log_notification(endpoint, "email", notif_type, "failed", message, str(exc))


async def send_discord(webhook_url: str, message: str, endpoint: Endpoint, result: dict, notif_type: str) -> None:
    color = 0xFF0000 if notif_type == "alert" else 0x00FF00
    fields = [
        {"name": "Endpoint", "value": endpoint.name, "inline": True},
        {"name": "URL", "value": endpoint.url, "inline": True},
        {"name": "Status", "value": result["status"], "inline": True},
    ]
    if notif_type == "alert":
        fields.append({"name": "Consecutive Failures", "value": str(endpoint.consecutive_failures), "inline": True})

    payload = {
        "embeds": [{
            "title": message,
            "color": color,
            "fields": fields,
            "timestamp": datetime.now(timezone.utc).isoformat(),
        }]
    }

    try:
        async with httpx.AsyncClient(timeout=10) as client:
            await client.post(webhook_url, json=payload)
        await log_notification(endpoint, "discord", notif_type, "sent", message)
    except Exception as exc:
        await log_notification(endpoint, "discord", notif_type, "failed", message, str(exc))


async def send_webhook(url: str, endpoint: Endpoint, result: dict, notif_type: str) -> None:
    payload = {
        "endpoint": endpoint.name,
        "url": endpoint.url,
        "status": result["status"],
        "consecutive_failures": endpoint.consecutive_failures,
        "checked_at": datetime.now(timezone.utc).isoformat(),
        "type": notif_type,
    }
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            await client.post(url, json=payload)
        await log_notification(endpoint, "webhook", notif_type, "sent", str(payload))
    except Exception as exc:
        await log_notification(endpoint, "webhook", notif_type, "failed", str(payload), str(exc))


async def log_notification(
    endpoint: Endpoint, channel: str, notif_type: str, status: str, message: str, error: str | None = None
) -> None:
    await Notification(
        endpoint_id=endpoint.id,
        channel=channel,
        type=notif_type,
        status=status,
        message=message,
        error=error,
    ).insert()
